### ISSUE TITLE
将DataLayer直接作为evaluator的输入出core

### ISSUE THREAD

**Issue by pengli09** [PaddlePaddle/Paddle#1623]
将DataLayer直接作为evaluator的输入时会出core：疑似DataLayer的forward()函数调用时机与其他layer不同，而evaluator并不会主动调用input的forward()函数，因此如果直接将DataLayer作为evaluator的输入，DataLayer的value和ids都将是空指针，导致出core。如下为可重现将问题的最小配置。

```
from paddle.trainer_config_helpers import *

dim = 4

define_py_data_sources2(
    train_list='train.list',
    test_list=None,
    module='data',
    obj='process',)

input = data_layer(name='input', size=dim)
label = data_layer(name='label', size=dim)
prob = fc_layer(input=input, size=4, act=SoftmaxActivation(), bias_attr=False)
classification_error_evaluator(input=input, label=label) # core here

outputs(sum_cost(prob))

settings(learning_method = AdamOptimizer(),
         batch_size = 5,
         learning_rate = 0)
```

**Comment by hedaoyuan** [comment]
这个错误的原因是在构造neural network的时候label被优化掉了。将`outputs(sum_cost(prob))`修改成`outputs(sum_cost(prob), label)`可以避免这个core。

**Comment by pengli09** [comment]
虽然用你提供的方法可以避免这个core，但是`label`在`classification_error_evaluator()`里面是用到了的，被优化掉不太合理也不符合直观，是否还是把这个bug修复一下呢？

**Comment by hedaoyuan** [comment]
这个会作为Bug看一下；不过当前的逻辑是evaluator并不是neural network的一部分，所以上面代码看起来是classification_error_evaluator里面用到了label，但实际上paddle在解析构造neural netwrok的时候是跳过这个语句的。
@reyoung paddle.v2里面对evaluator的处理也是一样的逻辑是吗？

**Comment by pengli09** [comment]
@reyoung @hedaoyuan 另外顺便提一下，如果将data_layer直接作为recurrent_layer_group的in_links，也会存在出core的情况，有可能和当前这个isssue是同一个问题导致的

**Comment by lcy-seso** [comment]
I close this issue due to inactivity. please feel free to reopen it.
